### DISCUSSION TITLE
Avoid wrapping single paragraphs in p tags?

### DISCUSSION THREAD

**Discussion by kokoshneta** [thephpleague/commonmark#812] [Q&A] [answered]
I would like to use Laravel’s `Str::markdown()` (which uses this library’s GithubFlavoredMarkdownConverter) to pull in translated text snippets for a multi-language site.

Most of these snippets are very short – just a few words or a single sentence – but some are longer, consisting of several paragraphs. In pretty much all cases where the snippet has no paragraph breaks, I would like to get back a converted string that’s _not_ wrapped in `<p>` tags, because these are meant to be used inline, as part of paragraphs, etc. Multi-paragraph snippets should still be properly paragraphed, though.

So for example, I would like to for these snippets:

```
Address
---
Categories
---
You **last** visited on
---
Lorem ipsum dolor sit amet, consectetur adipiscing elit.

Ut enim ad minim veniam, quis nostrud exercitation.
```

– to return this HTML:

```
Address
---
You <b>last</b> visited on
---
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p>

<p>Ut enim ad minim veniam, quis nostrud exercitation.</p>
```

I’ve searched through the documentation, particularly of course the configuration part, but I can’t find any mention of a setting or function that would provide this kind of functionality. It’s sort of like inline parsing, except that if the snippet explicitly contains block-level markdown (headers, paragraphs), those should be included.

Is it possible to do this?

**Comment by colinodell** [comment, marked as answer]
Hey there! There's no built-in functionality to enable that behavior.  You could probably accomplish this with by decorating the built-in `ParagraphRenderer` with one that immediately returns just the child inlines if it's the only paragraph.  Something like this (untested):

```php
final class AvoidParagraphTagsIfOnlyParagraphRenderer implements NodeRendererInterface
{
    private ParagraphRenderer $normalRenderer;

    public function __construct(ParagraphRenderer $normalRenderer)
    {
        $this->normalRenderer = $normalRenderer;
    }

    public function render(Node $node, ChildNodeRendererInterface $childRenderer): string
    {
        // Is this paragraph the only block in the document?
        if ($node->parent() instanceof Document && $node->previous() === null && $node->next() === null) {
            // Yes - so let's render it's children without the usual wrapping <p> tags
            return $childRenderer->renderNodes($node->children()) . $childRenderer->getBlockSeparator();
        }

        // Otherwise, render as usual
        return $this->normalRenderer->render($node, $childRenderer);
    }
}
```

You can then hook it up to your `Environment` like this:

```php
$environment->addRenderer(Paragraph::class, new AvoidParagraphTagsIfOnlyParagraphRenderer(new ParagraphRenderer()));
```

I haven't actually tested that code but it should work :)

**Reply by kokoshneta** [reply]
Thanks for confirming that!

I was hoping to use Laravel’s built-in `Str` helper class, which allows me to never actually load an environment or interact directly with Commonmark at all – and I’ve since realised that there are actually also some one-line snippets that _should_ be wrapped in paragraph tags. So overall, I think it might be easier if I just call the `Str::markdown()` function directly on those snippets that require it and leave it out where it’s not needed.
